File: frontend/src/types/scholarship.ts
// Types mirroring backend schemas/scholarship.py.

export interface Scholarship {
  id: string
  name: string
  slug?: string | null
  description?: string | null
  provider_name: string
  is_government: boolean
  amount?: number | null
  eligibility_criteria?: string | null
  income_criteria?: string | null
  deadline?: string | null
  documents?: string | null
  application_procedure?: string | null
  official_application_url?: string | null
  status: "active" | "expired" | "draft"
  states: string[]
  course_names: string[]
  source_name?: string | null
  source_url?: string | null
  verification_status?: string | null
  last_verified_at?: string | null
  created_at: string
  updated_at: string
}

export interface ScholarshipDetail extends Scholarship {
  course_ids: string[]
}

export interface ScholarshipFacets {
  total: number
  states: { label: string; count: number }[]
  courses: { label: string; count: number }[]
  statuses: { label: string; count: number }[]
}

export type ScholarshipListParams = {
  page?: number
  size?: number
  search?: string
  course?: string
  state?: string
  govt?: boolean
  status?: string
  upcoming?: boolean
  min_amount?: number
}

export interface ScholarshipCreateInput {
  name: string
  description?: string
  provider_name: string
  is_government?: boolean
  amount?: number
  eligibility_criteria?: string
  income_criteria?: string
  deadline?: string
  documents?: string
  application_procedure?: string
  official_application_url?: string
  status?: string
  states?: string[]
  course_ids?: string[]
}

File: backend/app/schemas/scholarship.py
from datetime import date, datetime
from typing import List, Optional

from pydantic import BaseModel, ConfigDict, Field, UUID4, field_validator


SCHOLARSHIP_STATUSES = ("active", "expired", "draft")
VERIFICATION_STATUSES = ("unverified", "pending", "verified")


class ScholarshipBase(BaseModel):
    name: str = Field(..., min_length=2, max_length=255)
    description: Optional[str] = Field(None, max_length=5000)
    provider_name: str = Field(..., min_length=2, max_length=255)
    is_government: bool = False
    amount: Optional[float] = Field(None, ge=0)
    eligibility_criteria: Optional[str] = Field(None, max_length=10000)
    income_criteria: Optional[str] = Field(None, max_length=2000)
    deadline: Optional[date] = None
    documents: Optional[str] = Field(None, max_length=5000)
    application_procedure: Optional[str] = Field(None, max_length=5000)
    official_application_url: Optional[str] = Field(None, max_length=500)
    status: str = "active"
    states: List[str] = []
    course_ids: List[UUID4] = []

    @field_validator("status")
    @classmethod
    def _validate_status(cls, v: str) -> str:
        stripped = v.strip().lower()
        if stripped not in SCHOLARSHIP_STATUSES:
            raise ValueError(f"status must be one of {sorted(SCHOLARSHIP_STATUSES)}")
        return stripped

    @field_validator("official_application_url")
    @classmethod
    def _validate_url(cls, v: Optional[str]) -> Optional[str]:
        if v is None or not v.strip():
            return None
        url = v.strip()
        if not (url.startswith("http://") or url.startswith("https://")):
            raise ValueError("official_application_url must start with http:// or https://")
        return url


class ScholarshipCreate(ScholarshipBase):
    pass


class ScholarshipUpdate(ScholarshipBase):
    pass


class ScholarshipPublishUpdate(BaseModel):
    status: str


class ScholarshipVerifyUpdate(BaseModel):
    verification_status: Optional[str] = None
    last_verified_at: Optional[datetime] = None

    @field_validator("verification_status")
    @classmethod
    def _validate(cls, v: Optional[str]) -> Optional[str]:
        if v is None:
            return v
        stripped = v.strip().lower()
        if stripped not in VERIFICATION_STATUSES:
            raise ValueError(f"verification_status must be one of {sorted(VERIFICATION_STATUSES)}")
        return stripped


class ScholarshipRead(BaseModel):
    id: UUID4
    name: str
    slug: Optional[str] = None
    description: Optional[str] = None
    provider_name: str
    is_government: bool = False
    amount: Optional[float] = None
    eligibility_criteria: Optional[str] = None
    income_criteria: Optional[str] = None
    deadline: Optional[date] = None
    documents: Optional[str] = None
    application_procedure: Optional[str] = None
    official_application_url: Optional[str] = None
    status: str = "active"
    states: List[str] = []
    course_names: List[str] = []
    source_name: Optional[str] = None
    source_url: Optional[str] = None
    verification_status: Optional[str] = None
    last_verified_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class ScholarshipDetail(ScholarshipRead):
    course_ids: List[UUID4] = []


class ScholarshipFacetBucket(BaseModel):
    label: str
    count: int


class ScholarshipFacets(BaseModel):
    total: int = 0
    states: List[ScholarshipFacetBucket] = []
    courses: List[ScholarshipFacetBucket] = []
    statuses: List[ScholarshipFacetBucket] = []